http-support/src/main/java/org/eclipse/sw360/antenna/http/impl/RequestBuilderImpl.java

```diff
@@ -0,0 +1,174 @@
+import org.eclipse.sw360.antenna.http.api.RequestBuilder;
+import org.eclipse.sw360.antenna.http.api.RequestProducer;
+
+import java.nio.file.Path;
+
+/**
+ * <p>
+ * Implementation of the {@code RequestBuilder} interface based on the builder
+ * class of OkHttpClient.
+ * </p>
+ */
+class RequestBuilderImpl implements RequestBuilder {
+    /**
+     * The mapper for doing JSON serialization.
+     */
+    private final ObjectMapper mapper;
+
+    /**
+     * The internal builder to delegate method calls to.
+     */
+    private final Request.Builder requestBuilder;
+
+    /**
+     * A builder for adding headers to the request.
+     */
+    private final Headers.Builder headersBuilder;
+
+    /**
+     * The name of the HTTP method to be invoked.
+     */
+    private String httpMethod;
+
+    /**
+     * The body of the request.
+     */
+    private RequestBody body;
+
+    /**
+     * Stores the name of a file to be uploaded. This field is only defined if
+     * a body of type file has been set.
+     */
+    private String fileName;
+
+    /**
+     * A builder for constructing a multi-part request. This is used only if
+     * the {@code bodyPart()} method is invoked.
+     */
+    private MultipartBody.Builder multipartBuilder;
+
+    /**
+     * Creates a new instance of {@code RequestBuilderImpl} to build a new
+     * request.
+     *
+     * @param mapper the JSON object mapper
+     */
+    public RequestBuilderImpl(ObjectMapper mapper) {
+        this.mapper = mapper;
+        requestBuilder = new Request.Builder();
+        headersBuilder = new Headers.Builder();
+        httpMethod = Method.GET.name();
+    }
+
+    @Override
+    public RequestBuilder method(Method method) {
+        httpMethod = method.name();
+        return this;
+    }
+
+    @Override
+    public RequestBuilder uri(String uri) {
+        requestBuilder.url(uri);
+        return this;
+    }
+
+    @Override
+    public RequestBuilder header(String name, String value) {
+        headersBuilder.add(name, value);
+        return this;
+    }
+
+    @Override
+    public RequestBuilder bodyString(String str, String mediaType) {
+        body = RequestBody.create(str, MediaType.parse(mediaType));
+        return this;
+    }
+
+    @Override
+    public RequestBuilder bodyFile(Path path, String mediaType) {
+        body = RequestBody.create(path.toFile(), MediaType.parse(mediaType));
+        Path fileNamePath = path.getFileName();
+        fileName = (fileNamePath != null) ? fileNamePath.toString() : null;
+        return this;
+    }
+
+    @Override
+    public RequestBuilder bodyJson(Object payload) {
+        try {
+            return bodyString(mapper.writeValueAsString(payload), "application/json");
+        } catch (JsonProcessingException e) {
+            throw new HttpExecutionException(e);
+        }
+    }
+
+    @Override
+    public RequestBuilder bodyPart(String name, RequestProducer partProducer) {
```

Comment:
I see that is where the multi part builder is used. This looks a bit ugly, it is a lot of complexity for calling a single method, or do I get something wrong here. Is it to fulfill the produceRequest interface?

Reply:
In Antenna there is the use case to generate multipart requests for file uploads. Such requests have a part with the binary content of the file and one JSON part with meta data. RequestBuilder needs to support such requests, and the idea was to do this via the _bodyPart()_ method that can simply be called multiple times for each part to be added - rather than calling one of the other methods to set a body.

There is actually a bigger overlap between a builder for a request part and one for a whole request. In both cases you can set headers and the body in different flavours (text, file, JSON).  The builder for a request part does not support setting a request URI, a method, and another body part though; therefore, these methods throw UnsupportedOperationExceptions in this implementation.

I thought I could be lazy and reuse the same interfaces in both cases due to the overlap in the properties that can be defined. But Stephanie commented on this as well; so I will think of an alternative.

Reply:
Thanks, the code is quite hard to understand as it is.

Reply:
The builder interfaces have been reworked. The bodies of regular requests and parts of a multipart request are now defined using a new RequestBodyBuilder interface. MultiPartRequestBuilder is no longer needed. I hope this is a clearer design.